ktor-server/jvm/src/io/ktor/server/application/plugins/api/KtorApplicationPlugin.kt

```diff
@@ -0,0 +1,480 @@
+ */
+public fun <A : Pipeline<*, ApplicationCall>, ConfigurationT : Any> A.plugin(
+    plugin: ApplicationInstallablePlugin<ConfigurationT>
+): KtorApplicationPlugin<ConfigurationT> {
+    return attributes[pluginRegistryKey].getOrNull(plugin.key)
+        ?: throw MissingApplicationPluginException(plugin.key)
+}
+
+internal fun <A : Pipeline<*, ApplicationCall>> A.findInterceptionsHolder(
+    plugin: ApplicationInstallablePlugin<*>
+): InterceptionsHolder {
+    return attributes[pluginRegistryKey].getOrNull(plugin.key)
+        ?: throw MissingApplicationPluginException(plugin.key)
+}
+
+public abstract class ApplicationInstallablePlugin<Configuration : Any>(public val name: String) :
+    ApplicationPlugin<ApplicationCallPipeline, Configuration, KtorApplicationPlugin<Configuration>>
+
+internal typealias PipelineHandler = (Pipeline<*, ApplicationCall>) -> Unit
+
+/**
+ * A plugin for Ktor that embeds into the HTTP pipeline and extends functionality of Ktor framework.
+ * */
+public abstract class KtorApplicationPlugin<Configuration : Any> private constructor(
+    installablePlugin: ApplicationInstallablePlugin<Configuration>
+) : PluginContext,
+    InterceptionsHolder {
+    protected abstract val pipeline: ApplicationCallPipeline
+
+    public abstract val pluginConfig: Configuration
+
+    public override val name: String = installablePlugin.name
+
+    public val environment: ApplicationEnvironment? get() = pipeline.environment
+    public val configuration: ApplicationConfig? get() = environment?.config
+
+    override val key: AttributeKey<KtorApplicationPlugin<Configuration>> = installablePlugin.key
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
+    override val fallbackInterceptions: MutableList<CallInterception> = mutableListOf()
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
+    override val callInterceptions: MutableList<CallInterception> = mutableListOf()
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
+    override val monitoringInterceptions: MutableList<CallInterception> = mutableListOf()
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
+    override val beforeReceiveInterceptions: MutableList<ReceiveInterception> = mutableListOf()
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
+    override val onReceiveInterceptions: MutableList<ReceiveInterception> = mutableListOf()
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
+    override val beforeResponseInterceptions: MutableList<ResponseInterception> = mutableListOf()
+
+    @Deprecated(level = DeprecationLevel.WARNING, message = "Please, do not use this field in your new plugin")
```

Comment:
Deprecation message should be reviewed @hhariri

Reply:
It is not needed anymore once we drop compatibility classes from public api